zulip_bots/zulip_bots/lib.py

```diff
@@ -137,8 +137,54 @@ def state(self, default):
         yield new_state
         self.set_state(new_state)
 
-def run_message_handler_for_bot(lib_module, quiet, config_file, bot_name):
-    # type: (Any, bool, str) -> Any
+def extract_query_without_mention(message, at_mention_bot_name):
+    # type: (Dict[str, Any], str) -> str
+    """
+    If the bot is the first @mention in the message, then this function returns
+    the message with the bot's @mention removed.  Otherwise, it returns None.
+    """
+    bot_mention = r'^@(\*\*{0}\*\*)'.format(at_mention_bot_name)
+    start_with_mention = re.compile(bot_mention).match(message['content'])
+    if start_with_mention is None:
+        return None
+    query_without_mention = message['content'][len(start_with_mention.group()):]
+    return query_without_mention.lstrip()
+
+def is_private(message, at_mention_bot_name):
+    # type: (Dict[str, Any], str) -> bool
+    # bot will not reply if the sender name is the same as the bot name
+    # to prevent infinite loop
+    if message['type'] == 'private':
+        return at_mention_bot_name != message['sender_full_name']
+    return False
+
+def initialize_config_bot(message_handler, bot_handler):
+    # type: (Any) -> None
+    if hasattr(message_handler, 'initialize'):
+        message_handler.initialize(bot_handler=bot_handler)
+
+def get_message_content_if_bot_is_called(message, at_mention_bot_name):
+    # type: (Dict[str, Any], str) -> Any
+    #
+    # is_mentioned is true if the bot is mentioned at ANY position (not necessarily
+    # the first @mention in the message).
+    is_mentioned = message['is_mentioned']
+    is_private_message = is_private(message=message, at_mention_bot_name=at_mention_bot_name)
+
+    # Strip at-mention botname from the message
+    if is_mentioned:
+        # message['content'] will be None when the bot's @-mention is not at the beginning.
+        # In that case, the message shall not be handled.
+        message['content'] = extract_query_without_mention(message=message, at_mention_bot_name=at_mention_bot_name)
+        if message['content'] is None:
```

Comment:
Leaving empty messages unhandled for now, as I saw a PR regarding empty messages by @neiljp.

We can remove merge conflicts as and when any of the PR gets merged. :)